2019-08-03-q-and-as-about-nix-for-haskellers.md

```diff
@@ -0,0 +1,692 @@
+title = Questions and Answers about Nix for Haskellers
+tags = haskell, nixos
+summary = Overview of Nix for Haskell development
+draft = true
+
+------------------------------------------------------
+
+Last week Dmitrii Kovanikov ([@ChShersh](https://twitter.com/ChShersh)) posted
+a [request](https://twitter.com/ChShersh/status/1154782785967542272?s=20) on
+Twitter asking if anyone had the time to answer some questions about how to
+understand Nix as it applies to Haskell programming.
+
+Dmitrii is working on a scaffolding tool for Haskell projects called
+[summoner](https://github.com/kowainik/summoner).  There is an
+[open PR](https://github.com/kowainik/summoner/pull/283) adding Nix support.
+However, Dmitrii doesn't have any Nix experience, so he'd like to learn more
+about using Nix for Haskell development before tackling the PR.
+
+I've been using Nix for about a year and half, both at work and on my own
+projects.  I think there is a big lack of documentation surrounding Nix and
+Haskell, so I told him I'd be able to answer any questions he had.
+
+He wrote me back with a _great_ list of questions.  I thought it would be an
+injustice to the Haskell community if I didn't share my answers, so I decided
+to turn his questions and my answers into an article.
+
+This article is mainly for Haskellers that have heard about Nix (on Twitter,
+Reddit, YouTube, etc), know about its advantages, but haven't actually played
+around with it.  It is mainly a **high-level overview** of how Nix works with
+Haskell.  It is not a hands-on tutorial or walk-through.
+
+The first section of this article is an intro to Nix, followed by answers
+to each of Dmitrii's questions.[^7]
+
+## Intro
+
+You may have heard that Nix is complicated and hard to learn.  In my
+experience, this is true.
+
+It is similar to Haskell.  The learning curve is steep, but once you get the
+hang of it, it is very powerful and flexible.
+
+*What makes it so hard?*
+
+To start off with, Nix refers to two things:
+
+1. A programming language for defining _derivations_
+
+2. A set of command line tools for actually building and installing _derivations_
+
+What is a _derivation_?  You can think of it as something similar
+to a "package" in a distribution like Debian or Fedora.
+
+The Nix language provides a way of defining derivations.  The following code
+shows an example of defining a derivation for the GNU
+[hello](https://www.gnu.org/software/hello/hello.html) package[^1]:
+
+```nix
+stdenv.mkDerivation {
+  name = "hello-2.1.1";
+  builder = ./builder.sh;
+  src = fetchurl {
+    url = ftp://ftp.nluug.nl/pub/gnu/hello/hello-2.1.1.tar.gz;
+    sha256 = "1md7jsfd8pa45z73bz1kszpp01yw6x5ljkjk2hx7wl800any6465";
```

Comment:
Yeah, I have a problem with this example. Personally, I wouldn't assume that people who heard about `nix` are familiar with its syntax. There's a great blog post about nix syntax:

* https://medium.com/@MrJamesFisher/nix-by-example-a0063a1a4c55

And I don't propose to vendor full nix syntax tutorial here. But I think that probably it can be a good idea to give 3 small examples of nix language (with results shown in repl?) in the increasing difficulty. Something like:

1. Simple string concatenation (probably with interpolation)
2. Example of a record
3. Example of a function call

From this example it's not clear that `mkDerivation` is a function. And that it's actually called on a record. And that this record is extensible and has some default arguments. Also, what is `stdenv` and where it comes from :thinking: